### ISSUE TITLE
I/O operation on closed file

### ISSUE THREAD

**Issue by serhiy-storchaka** [psf/black#1664]
Running **Black** 20.8b1 with the `--diff` option on a tree (> 100 files) produces errors like
```
error: cannot format tests/api/test_core.py: I/O operation on closed file
```
Running `black --diff` with a single file does not produce this error.

Linux/Python 3.7.8+

**Comment by jpgrayson** [comment]
The problematic behavior only happens when the optional `colorama` package is installed.

When wrapping the output stream with colorama, it is assumed that colorama returns a wrapped stream without its own `detach()` method, but a wrapped stream is only returned on Windows platforms. The problem is that when no wrapper is affected, the code still unconditionally reassigns the original stream's `detach()` method to be a no-op.

I'll post a PR for this shortly.

**Comment by ichard26** [comment]
Running `black -v --diff -l 79` over [some typeshed stubs](https://github.com/python/typeshed/tree/59586848c97075b13271f4ca86fe0593c7fe19ec/stdlib/2):

Before 4d71d74a442ccc6c309a0667147997c1eeb755fd:
```console
2 files would be reformatted, 56 files would be left unchanged, 86 files would fail to reformat.
```

After 4d71d74a442ccc6c309a0667147997c1eeb755fd:
```console
88 files would be reformatted, 56 files would be left unchanged.
```

<details>
<summary><b>My environment</b></summary>

- Python version: `CPython 3.8.5`
- Black version: look above
- OS version: `Ubuntu 20.04.01 LTS`

</details>

I'll have to check that the patch also works on Windows but that will take a while -  hopefully I don't have mess around with Windows Update :-)

**Comment by ichard26** [comment]
> I'll have to check that the patch also works on Windows but that will take a while - hopefully I don't have mess around with Windows Update :-)

I procrastinated for way too long on getting back if this patch works on Windows or not but the answer is "yes it does" :D

**Comment by ichard26** [comment]
This was fixed quite a while ago, although it was only released [in 21.4b0](https://black.readthedocs.io/en/latest/change_log.html#id6).
